### PR TITLE
use pyav to encode straight to h264, remove docker ffmpeg dep.

### PR DESCRIPTION
**Pull request by orm011**
…ove ffmpeg docker dep.

The larger goal is to move toward not requiring docker. there's a separate, orthogonal pull request that enables this for postgres.

In this case, pyav is a pip installable python binding for libav.
As a side effect of using this, you also dont need to render the video twice to get around the licensing issue you mentioned.

videos look fine.

### CODE REVIEW

File: pixeltable/functions/__init__.py [hunk 1]
```diff
@@ -102,31 +102,34 @@ def value(self) -> float:
 
 class VideoAggregator:
     def __init__(self):
-        self.video_writer = None
-        self.size = None
+        '''follows https://pyav.org/docs/develop/cookbook/numpy.html#generating-video'''
+        self.container = None
+        self.stream = None
+        self.fps = 24
```
Comment (mkornacker) [inline pixeltable/functions/__init__.py]: We're using fps=25 at the moment, so let's stick with that.

File: pixeltable/functions/__init__.py [hunk 2]
```diff
@@ -102,31 +102,34 @@ def value(self) -> float:
 
 class VideoAggregator:
     def __init__(self):
-        self.video_writer = None
-        self.size = None
+        '''follows https://pyav.org/docs/develop/cookbook/numpy.html#generating-video'''
```
Comment (mkornacker) [inline pixeltable/functions/__init__.py]: """

File: pixeltable/functions/__init__.py [hunk 3]
```diff
@@ -102,31 +102,34 @@ def value(self) -> float:
...
+        self.container = None
```
Comment (mkornacker) [inline pixeltable/functions/__init__.py]: missing typing.

File: pixeltable/functions/__init__.py [hunk 4]
```diff
@@ -102,31 +102,34 @@ def value(self) -> float:
...
-            self.video_writer = cv2.VideoWriter(str(self.tmp_file), cv2.VideoWriter_fourcc(*'mp4v'), 25, self.size)
-
-        frame_array = np.array(frame)
-        frame_array = cv2.cvtColor(frame_array, cv2.COLOR_RGB2BGR)
-        self.video_writer.write(frame_array)
-
+            self.container = av.open(str(self.out_file), mode="w")
+            self.stream = self.container.add_stream("mpeg4", rate=self.fps)
+            self.stream.width = frame.width
+            self.stream.height = frame.height
```
Comment (mkornacker) [inline pixeltable/functions/__init__.py:122]: do you need stream.pix_fmt = "yuv420p" anywhere?

**Review by mkornacker — COMMENTED** [review]
Have you run the tutorial notebooks by hand to see if everything works?

File: pixeltable/functions/__init__.py (hunk 4 above)
Comment (orm011) [reply pixeltable/functions/__init__.py:122]: Doesn't seem like it. The generated videos seem to have yuv420p either way.

**Comment by orm011** [comment]
> Have you run the tutorial notebooks by hand to see if everything works?

I ran the 'Comparing Object Detection Models for Video.ipynb'. 
That one does work (the generated video is available and looks ok).

There is currently a bug in master though, where the .show() does not seem to include the url for the video in the video tag. 
That's a separate issue though

File: pixeltable/functions/__init__.py [hunk 5]
```diff
@@ -102,29 +103,32 @@ def value(self) -> float:
 
 class VideoAggregator:
     def __init__(self):
-        self.video_writer = None
-        self.size = None
+        """follows https://pyav.org/docs/develop/cookbook/numpy.html#generating-video"""
+        self.container : av.container.OutputContainer = None
```
Comment (mkornacker) [inline pixeltable/functions/__init__.py]: self.container: Optional[...] = None

File: pixeltable/functions/__init__.py [hunk 6]
```diff
@@ -102,29 +103,32 @@ def value(self) -> float:
...
+        self.stream : av.stream.Stream = None
```
Comment (mkornacker) [inline pixeltable/functions/__init__.py]: same

**Review by mkornacker — APPROVED** [review]

**Pull request merged**